File: 1046-last-stone-weight/1046-last-stone-weight.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        auto compare = [](int a,int b) {return a < b;};
        make_heap(stones.begin(), stones.end(), compare);
        
        int s1,s2, cursize = stones.size();
        while (cursize > 1) {
            pop_heap(stones.begin(), stones.end(), compare);
            s1 = stones[cursize-1];
            --cursize;
            stones.pop_back();
            pop_heap(stones.begin(), stones.end(), compare);
            s2 = stones[cursize-1];
            --cursize;
            stones.pop_back();
            if (s1!=s2) {
                stones.push_back(s1-s2);
                push_heap(stones.begin(), stones.end(), compare);
                ++cursize;
            }
        }
        
        if (cursize == 0) return 0;
        return stones[0];
    }
};

File: 1046-last-stone-weight/1046-last-stone-weight.java
class Solution {
    public int lastStoneWeight(int[] stones) {
        Queue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        
        for (int stone : stones) {
            pq.add(stone);
        }
        
        while (pq.size() > 1) {
            int y = pq.poll();
            int x = pq.poll();
            
            if (x==y) {
                continue;
            } else {
                pq.add(y-x);
            }
        }
        
        return (pq.size()==0)?0:pq.poll();
    }
}